File: independent_campaign/experiments/torus_sq.py
"""g(q) = max square-corner-free subset of the TORUS (Z_q)^2.

Why this is the decisive quantity.

TENSOR LEMMA (proved, no carry conditions needed).  Let T subset (Z_q)^2 be
square-corner-free on the torus.  Put S = { sum_j d_j q^j : d_j in T } subset [q^d]^2.
Then S is square-corner-free and |S| = |T|^d.
  Proof.  Let b,u,v in S with v = i*u + (1-i)*b, u != b.  Reducing mod q gives
  v_0 = i*u_0 + (1-i)*b_0 with b_0,u_0,v_0 in T, which is a torus square corner unless
  u_0 = b_0; so u_0 = b_0, hence v_0 = b_0 as well.  Then w = u-b = 0 mod q, so
  w = q w', and dividing the equation by q leaves the same configuration on the
  d-1 higher digits.  Induction on d.  QED

Hence   Q_SQ(n) >= n^{log g(q) / log q}   for every q,   so the exponent
        gamma = lim log Q_SQ(n)/log n   satisfies   gamma >= sup_q log g(q)/log q.

A LINE {(t, a t) : t in Z_q} with 1 + a^2 invertible mod q is torus square-corner-free
(w and i*w both on the line forces (1+a^2) w = 0), so g(q) >= q always, giving
gamma >= 1 -- which is only the trivial bound.  The decisive question is therefore

        IS g(q) > q FOR SOME q ?

If yes, square-corner-free sets are provably SUPERLINEAR and the relaxation cannot
prove C(n) = O(n^{1+o(1)}).  If g(q) = q for all q, that is strong evidence the
relaxation is tight at n^{1+o(1)} and the route survives.

Exhaustive branch and bound, exact integer arithmetic on the torus.
"""
import sys
from time import time


def build(q):
    """cells, and for each ordered pair (b,u) the third cell v = b + i*(u-b)."""
    N = q * q
    idx = lambda x, y: (x % q) * q + (y % q)
    third = [[0] * N for _ in range(N)]
    for bx in range(q):
        for by in range(q):
            b = idx(bx, by)
            for ux in range(q):
                for uy in range(q):
                    u = idx(ux, uy)
                    wx, wy = ux - bx, uy - by
                    third[b][u] = idx(bx - wy, by + wx)
    return N, third


def solve(q, verbose=True, best_known=None):
    """exact max square-corner-free subset of (Z_q)^2"""
    N, third = build(q)
    # forb[b][u] = bitmask of cells that become illegal once both b and u are chosen.
    # A square corner needs an apex and two legs; with S known, adding p can create a
    # corner in three ways (p as apex, p as first leg, p as second leg).  We handle it
    # by the direct incremental test below instead of precomputed masks, which keeps
    # the code obviously correct.
    best = [0 if best_known is None else best_known]
    bestset = [None]
    nodes = [0]

    def addable(S, p):
        """is S+{p} still square-corner-free?  Checks every triple containing p."""
        Sset = S[1]
        for b in S[0]:
            # p as a leg with apex b, other leg must not be present
            if third[b][p] in Sset:
                return False
            # p as apex
            if third[p][b] in Sset:
                return False
            # b as a leg with apex p handled by third[p][b]; p as the SECOND leg:
            # need b' with third[b'][b] == p  -> scan
        for b in S[0]:
            for c in S[0]:
                if third[b][c] == p:
                    return False
        return True

    order = list(range(N))

    def rec(S, cand):
        nodes[0] += 1
        if len(S[0]) + len(cand) <= best[0]:
            return
        if not cand:
            if len(S[0]) > best[0]:
                best[0] = len(S[0])
                bestset[0] = list(S[0])
            return
        if len(S[0]) > best[0]:
            best[0] = len(S[0])
            bestset[0] = list(S[0])
        for k, p in enumerate(cand):
            if len(S[0]) + len(cand) - k <= best[0]:
                return
            S[0].append(p)
            S[1].add(p)
            rest = [x for x in cand[k + 1:] if addable(S, x)]
            rec(S, rest)
            S[0].pop()
            S[1].discard(p)

    t0 = time()
    # symmetry: fix that cell 0 = (0,0) is in S (the problem is translation invariant
    # on the torus, so every nonempty solution has a translate containing (0,0))
    S = [[0], {0}]
    cand = [p for p in order[1:] if addable(S, p)]
    rec(S, cand)
    dt = time() - t0
    if verbose:
        print(f"q={q:3d}  g(q) = {best[0]:4d}   (q = {q})   "
              f"log g/log q = {__import__('math').log(best[0])/__import__('math').log(q):.5f}   "
              f"nodes={nodes[0]:,}  {dt:.1f}s")
        if best[0] > q:
            print(f"      *** g({q}) > {q}: SUPERLINEAR.  witness = "
                  f"{sorted((c//q, c%q) for c in bestset[0])}")
    return best[0], bestset[0]


if __name__ == "__main__":
    qs = [int(a) for a in sys.argv[1:]] or [2, 3, 4, 5, 6, 7, 8, 9]
    for q in qs:
        solve(q)

File: independent_campaign/experiments/torus_sq.c
/* g(q) = max square-corner-free subset of the torus (Z_q)^2, exhaustive.
 *
 * A square corner is {b, u, v} with v = b + i*(u-b), u != b, i*(x,y) = (-y,x).
 * For an unordered pair {a,b} the cells x completing a corner are exactly four:
 *    t(a,b) = a + i*(b-a)              (apex a)
 *    t(b,a) = b + i*(a-b)              (apex b)
 *    x with t(x,a) = b, i.e. x = (1-i)^{-1}(b - i a)     (apex x, legs a,b)
 *    x with t(x,b) = a, i.e. x = (1-i)^{-1}(a - i b)
 * (1-i) is the matrix [[1,1],[-1,1]], det 2, so it is invertible mod odd q; for even q
 * the last two cells are computed by an explicit table search instead.
 *
 * DFS over cells in increasing index order carrying a candidate bitset with the
 * invariant "every candidate is individually addable to S and has a larger index".
 * Bound: |S| + popcount(cand) <= best  =>  prune.  Translation invariance on the torus
 * lets us fix cell (0,0) in S without loss.
 *
 * usage: torus_sq q [start_best]
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#define WMAX 8                      /* 8*64 = 512 bits, enough for q <= 22 */
static int Q, N, NW;

typedef struct { uint64_t w[WMAX]; } BS;

static int  third[512][512];        /* third[b][u] = b + i*(u-b) */
static int  fourcell[512][512][8];  /* the <=6 completions of the pair {a,b} */
static int  nfour[512][512];
static int  pairbad[512][512];      /* {a,x} alone already infeasible (even q only) */

static inline int idx(int x, int y) { x %= Q; if (x < 0) x += Q; y %= Q; if (y < 0) y += Q; return x * Q + y; }

static inline void bs_clear(BS *a)              { memset(a->w, 0, sizeof(a->w)); }
static inline void bs_set(BS *a, int i)         { a->w[i >> 6] |= 1ULL << (i & 63); }
static inline int  bs_get(const BS *a, int i)   { return (a->w[i >> 6] >> (i & 63)) & 1ULL; }
static inline void bs_unset(BS *a, int i)       { a->w[i >> 6] &= ~(1ULL << (i & 63)); }
static inline int  bs_count(const BS *a) {
    int c = 0; for (int k = 0; k < NW; k++) c += __builtin_popcountll(a->w[k]); return c;
}

static long long nodes = 0;
static int best = 0, bestset[512], curset[512];

static void rec(int depth, BS cand)
{
    nodes++;
    if (depth > best) { best = depth; memcpy(bestset, curset, sizeof(int) * depth); bestset[depth] = -1;
                        printf("   .. new best %d\n", best); fflush(stdout); }
    int rem = bs_count(&cand);
    if (depth + rem <= best) return;

    for (int p = 0; p < N; p++) {
        if (!bs_get(&cand, p)) continue;
        /* remaining candidates from p on */
        if (depth + rem <= best) return;
        rem--;                                  /* p is consumed by this branch */
        BS nc = cand;
        for (int j = 0; j <= p; j++) bs_unset(&nc, j);   /* increasing index order */
        /* filter: kill everything that would complete a corner with p and some s in S */
        for (int x = 0; x < N; x++) if (pairbad[p][x]) bs_unset(&nc, x);
        for (int s = 0; s < depth; s++) {
            int a = curset[s];
            for (int k = 0; k < nfour[a][p]; k++) bs_unset(&nc, fourcell[a][p][k]);
        }
        curset[depth] = p;
        rec(depth + 1, nc);
    }
}

int main(int argc, char **argv)
{
    Q = atoi(argv[1]);
    N = Q * Q;
    NW = (N + 63) / 64;
    if (NW > WMAX) { printf("q too large\n"); return 1; }
    if (argc > 2) best = atoi(argv[2]);

    for (int bx = 0; bx < Q; bx++) for (int by = 0; by < Q; by++)
    for (int ux = 0; ux < Q; ux++) for (int uy = 0; uy < Q; uy++) {
        int wx = ux - bx, wy = uy - by;
        third[idx(bx,by)][idx(ux,uy)] = idx(bx - wy, by + wx);
    }
    /* fourcell: for the pair (a,b), all x with {a,b,x} a square corner */
    for (int a = 0; a < N; a++) for (int b = 0; b < N; b++) {
        if (a == b) { nfour[a][b] = 0; continue; }
        /* x completes the pair {a,b} iff SOME ordered pair (p,r) drawn from the three
         * cells has third[p][r] equal to the remaining cell.  All six are checked. */
        int m = 0;
        for (int x = 0; x < N; x++) {
            if (x == a || x == b) continue;
            if (third[a][b] == x || third[b][a] == x ||
                third[a][x] == b || third[x][a] == b ||
                third[b][x] == a || third[x][b] == a) {
                if (m < 8) fourcell[a][b][m++] = x;
            }
        }
        nfour[a][b] = m;
    }
    /* pair-level infeasibility: an ordered pair (a,x), a != x, whose forbidden third
     * cell IS x itself.  Only possible when (1-i) is a zero divisor, i.e. q even. */
    for (int a = 0; a < N; a++) for (int x = 0; x < N; x++) {
        pairbad[a][x] = (a != x) && (third[a][x] == x || third[x][a] == a ||
                                     third[a][x] == a || third[x][a] == x);
    }

    /* Symmetry.  The defining equation v = i*u + (1-i)*b is Z[i]-linear, so every map
     * z -> alpha*z + beta with alpha a unit of Z_q[i], and complex conjugation, sends
     * square-corner-free sets to square-corner-free sets.  Translations let us put
     * 0 in S; then (lex-min argument) the second element may be assumed minimal in its
     * orbit under Stab(0) = units together with conjugation. */
    int orbmin[512]; memset(orbmin, 0, sizeof(orbmin));
    int nunit = 0;
    for (int d = 1; d < N; d++) {
        int dx = d / Q, dy = d % Q, mn = d;
        for (int ax = 0; ax < Q; ax++) for (int ay = 0; ay < Q; ay++) {
            int nrm = (ax * ax + ay * ay) % Q, inv = 0;
            for (int t = 1; t < Q; t++) if (nrm * t % Q == 1) inv = 1;
            if (!inv) continue;                       /* alpha must be a unit */
            if (d == 1) nunit++;
            /* alpha * z  and  alpha * conj(z) */
            int m1 = idx(ax * dx - ay * dy, ax * dy + ay * dx);
            int m2 = idx(ax * dx + ay * dy, -ax * dy + ay * dx);
            if (m1 && m1 < mn) mn = m1;
            if (m2 && m2 < mn) mn = m2;
        }
        orbmin[d] = (mn == d);
    }
    int nroot = 0; for (int d = 1; d < N; d++) nroot += orbmin[d];
    printf("q=%d: %d units, %d orbit-minimal second points (of %d)\n",
           Q, nunit, nroot, N - 1);

    BS cand; bs_clear(&cand);
    for (int p = 1; p < N; p++) bs_set(&cand, p);
    for (int x = 0; x < N; x++) if (pairbad[0][x]) bs_unset(&cand, x);
    curset[0] = 0;
    if (getenv("NOSYM")) {
        rec(1, cand);
    } else {
        best = best > 1 ? best : 1;
        for (int d = 1; d < N; d++) {
            if (!orbmin[d] || !bs_get(&cand, d)) continue;
            BS nc = cand;
            for (int x = 0; x <= d; x++) bs_unset(&nc, x);
            for (int x = 0; x < N; x++) if (pairbad[d][x]) bs_unset(&nc, x);
            for (int k = 0; k < nfour[0][d]; k++) bs_unset(&nc, fourcell[0][d][k]);
            curset[0] = 0; curset[1] = d;
            rec(2, nc);
        }
    }

    printf("q=%d  g(q)=%d  nodes=%lld\n", Q, best, nodes);
    printf("witness:");
    for (int k = 0; k < best; k++) printf(" (%d,%d)", bestset[k] / Q, bestset[k] % Q);
    printf("\n");
    return 0;
}
